### PR TITLE
Ensure that DaemonSet respects termination

### LINKED ISSUE #50477: DaemonSet controller start new pod when the old pod is still in Terminating state
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
The DaemonSet controller will schedule a pod on a node when the old pod is still in Terminating state. This breaks the expectation that only one pod will be running at a time per node.

**What you expected to happen**:
The DaemonSet controller should not let two pods running at the same time on the same node.

**How to reproduce it (as minimally and precisely as possible)**:
Create a DaemonSet. Set terminationGracePeriodSeconds to be relatively long. Kill a pod. Watch a new pod start on the same node when the old pod is still terminating.

**Anything else we need to know?**:
Seems this behavior was introduced in https://github.com/kubernetes/kubernetes/pull/43077

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.2
- Cloud provider or hardware configuration**: GKE
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by kow3ns**
**What this PR does / why we need it**:
#43077 correctly prevents the DaemonSet controller from adopting deleted Pods, but, as pointed out in #50477, the controller now has no sensitivity to the termination lifecycle (i.e TerminationGracePeriodSeconds) of the Pods it creates. This PR attempts to balance the two. DaemonSet controller will now consider deleted Pods owned by a DaemonSet during creation, but it will not consider deleted Pods as targets for adoption.

fixes #50477

```release-note
Fix a problem of not respecting TerminationGracePeriodSeconds of the Pods created by DaemonSet controller.
```

### CODE REVIEW

**Comment by enisoc** [comment]
The purpose of #43077 was not to prevent adoption of terminating Pods. That was accomplished independently with #44150, which means you shouldn't need to make any changes to `getDaemonPods`.

**Comment by kow3ns** [comment]
@enisoc #43077 doesn't allow any Pods with a DeletionTimestamp to be considered with by ControllerRefManager. #44150 filters Pods with a DeletionTimestamp.
[This line](https://github.com/kubernetes/kubernetes/pull/51279/files#diff-3af9d435092399b4bf3ef61e0387eb36L758) makes it impossible for the DaemonSet Controller to be responsive to terminated Pods. If you're claiming [this](https://github.com/kubernetes/kubernetes/pull/51279/files#diff-3af9d435092399b4bf3ef61e0387eb36R758) is unnecessary, I can try removing it, but we most certainly need to change `getDaemonPods` to ensure that main loop actually sees terminated Pods.

**Comment by enisoc** [comment]
The logic added in #44150 only applies to orphans. `ClaimPods()` will already include terminating Pods in its result, as long as you own them:

https://github.com/kubernetes/kubernetes/blob/05e7f6d0737be8161469c9f78660e25122ed838b/pkg/controller/controller_ref_manager.go#L69-L80

So your change to add back `ownedAndDeleted` Pods is unnecessary. They are already in `claimed`.

**Comment by kow3ns** [comment]
I see so the code introduced in #43077 is completely made useless by #44150. After this PR ControllerRefManager does the right thing. That makes this PR shorter.

**Comment by enisoc** [comment]
Do we also need to bring back protection against double deletion?

https://github.com/kubernetes/kubernetes/pull/43077/files#diff-cb9055723bef8b239f9df77ac704ffacL556

Your PR is now basically a straight revert of #43077. I still think they intended that to have some effect unrelated to the problem of adopting terminating orphans (which they themselves fixed separately in #44150). We should make sure we understand what problem #43077 was trying to fix before we revert it.

cc @lukaszo @kargakis

File: pkg/controller/daemon/daemon_controller_test.go [hunk 1]
```diff
@@ -1816,12 +1831,6 @@ func TestGetNodesToDaemonPods(t *testing.T) {
 			newPod("non-matching-owned-0-", "node-0", simpleDaemonSetLabel2, ds),
 			newPod("non-matching-orphan-1-", "node-1", simpleDaemonSetLabel2, nil),
 			newPod("matching-owned-by-other-0-", "node-0", simpleDaemonSetLabel, ds2),
-			func() *v1.Pod {
```
Comment (enisoc) [inline pkg/controller/daemon/daemon_controller_test.go]: Maybe move this up to `wantedPods` so we are clear to future readers that this behavior is required.

Comment (kow3ns) [reply pkg/controller/daemon/daemon_controller_test.go]: SGTM

**Comment by enisoc** [comment]
@kow3ns It seems like the intent of #43077 was to omit terminating-but-owned Pods from the "available" count. However, for some reason no test case was added for such a Pod:

https://github.com/kubernetes/kubernetes/pull/43077/files#diff-7f51010c4c3f3050dcce5c9d916fc3eeR138

Could you add a test case for a terminating-but-owned Pod in that function? I expect it will fail now that you've reverted the change from #43077. To avoid regressing on the bug that was meant to fix, we may have to add a new, more targeted fix in `getUnavailableNumbers()`.

Also, you mentioned offline that you brought back the protection against double deletion. Did you not push that commit yet?

**Comment by janetkuo** [comment]
I suggest we revert #43077 but keep the test cases. If we want to count terminating pods as unavailable (what #43077 originally tried to fix), we should update below code to not increment `numberAvailable` when the pod is terminating:

https://github.com/kubernetes/kubernetes/blob/05e7f6d0737be8161469c9f78660e25122ed838b/pkg/controller/daemon/daemon_controller.go#L1033-L1035


I originally suggested that `IsPodAvailable` should count terminating pods as unavailable, but the decision made then was to let controllers to decide whether to filter out terminating pods or not:
https://github.com/kubernetes/kubernetes/pull/32771/files#diff-473b59d03b4c3dc229240ad9fb29c1edR84

We should probably revisit this function if we think it's a common requirement for all controllers to determine availability of pods.

**Comment by lukaszo** [comment]
The goal of #43077 was to make daemon controller behavior similar to replicaset controller where pods marked for deletion are also ignored. 
If the behavior is wrong then probably ReplicaSet controller also should be fixed.

**Comment by lukaszo** [comment]
>  However, for some reason no test case was added for such a Pod:
@enisoc it's handled here https://github.com/kubernetes/kubernetes/pull/43077/files#diff-0dc431d36cd3969089646d1bd4640d1eR1270

**Comment by janetkuo** [comment]
> The goal of #43077 was to make daemon controller behavior similar to replicaset controller where pods marked for deletion are also ignored.
If the behavior is wrong then probably ReplicaSet controller also should be fixed.

It is okay to see deleting pods as unavailable during a rolling update (when making sure `maxUnavailable` isn't violated). 
It is not okay to ignore deleting pods and try to create a new pod on the same node (#50477).

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -770,10 +771,6 @@ func (dsc *DaemonSetsController) getNodesToDaemonPods(ds *extensions.DaemonSet)
 	// Group Pods by Node name.
 	nodeToDaemonPods := make(map[string][]*v1.Pod)
 	for _, pod := range claimedPods {
-		// Skip terminating pods
-		if pod.DeletionTimestamp != nil {
-			continue
-		}
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Add these back to `dsc.manage()` (where it's [deleted](https://github.com/kubernetes/kubernetes/pull/43077/files#diff-cb9055723bef8b239f9df77ac704ffacL556))?

**Comment by kow3ns** [comment]
@janetkuo @enisoc @kargakis are we good with this?

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
Since the regression was introduced in 1.7, we should probably cherrypick this one into 1.7

**Pull request merged**